### ISSUE TITLE
Optimizing fontTools.ttLib.TTFont.getGlyphID

### ISSUE THREAD

**Issue by dscorbett** [fonttools/fonttools#1536]
I have a project with many thousands of glyphs, which translates to thousands of calls to `fontTools.ttLib.TTFont.getGlyphID` when calling `fontTools.ttLib.TTFont.save`. I was able to make it faster (17.66&nbsp;s to 4.236&nbsp;s of cumulative time) by adding the following to the beginning of the function.
```python
		try:
			return self._reverseGlyphOrderDict[glyphName]
		except (AttributeError, KeyError):
			pass
```

The function is slow because it checks that `_reverseGlyphOrderDict` and `glyphOrder` are up to date and consistent. My version skips those checks. Is that safe? I think it is, for my use case, because I have finished creating glyphs by the time the function is first called.

Would it be possible to add something like this optimization to fontTools? If `getGlyphID` is called because `save` was called, it seems like it could assume that `_reverseGlyphOrderDict` and `glyphOrder` are frozen for the duration of `save`.

**Comment by behdad** [comment]
> I was able to make it faster (17.66 s to 4.236 s of cumulative time) by adding the following to the beginning of the function.

That's super weird.  It shouldn't take that long no matter what.  Do we know which callsites are taking the most time?

**Comment by behdad** [comment]
How long does the save take in total?

**Comment by dscorbett** [comment]
I reran the build and have attached [the profiles](https://github.com/fonttools/fonttools/files/2959176/profiles.tar.gz). `save` took 50.375&nbsp;s in the slower build vs. 42.159&nbsp;s in the faster build. The most time-consuming callers of `getGlyphID` in the slower build were [`GlyphID.write`](https://github.com/fonttools/fonttools/blob/3.26.0/Lib/fontTools/ttLib/tables/otConverters.py#L281), [`Coverage.preWrite`](https://github.com/fonttools/fonttools/blob/3.26.0/Lib/fontTools/ttLib/tables/otTables.py#L487), and [`SingleSubst.preWrite`](https://github.com/fonttools/fonttools/blob/3.26.0/Lib/fontTools/ttLib/tables/otTables.py#L648).

**Comment by anthrotype** [comment]
@dscorbett How did you create those .profile files? They don't seem to be the output of `python -m cProfile -o ...` -- I can't visualize them or extract the data.

**Comment by dscorbett** [comment]
They are the output of `python -m cProfile -o ...` using Python 2.7. I can visualize them in SnakeViz and extract their data using [`pstats.Stats`](https://docs.python.org/2/library/profile.html#pstats.Stats).

**Comment by anthrotype** [comment]
ah ok, I was trying with python3.7 that's why it failed.
(you know we'll be dropping support for py27 soon-ish, in the coming months. I just have to find the time to do it)

**Comment by anthrotype** [comment]
i'm a bit surprised that calling getGlyphID is three times slower than directly looking up self._reverseGlyphOrderDict.
For the most common case of a glyphName which is already contained in self._reverseGlyphOrderDict, and the latter being in sync with self.glyphOrder, the extra work is basically:
1) one hasattr check (line 566),
2) one call to getGlyphOrder, which in turn simply returns self.glyphOrder (line 568),
3) one dict contains check (line 570)
4) one dict getitem call (line 602),
5) and one string comparison (line 603).

I wonder why we care about checking the consistency between glyphOrder and _reverseGlyphOrderDict every single time getGlyphID is used.
Would it be enough to just have `setGlyphOrder` method be the only supported way of modifying the self.glyphOrder, and have that method invalidate (i.e. remove the attribute) the self._reverseGlyphOrderDict cache? Then the `getGlyphID` method can simply assume that whenever a `_reverseGlyphOrderDict` attribute is present, it must be in sync with the glyphOrder, and do try get glyphName from it, except rebuilding the reverse order when missing.

If a client messes directly with `self.glyphOrder` or with `self._reversedGlyphOrderDict` it's not fontTools' fault.

**Comment by dscorbett** [comment]
It is three times slower because the function is doing so little (in the most common case) that even a few extra steps have a large relative impact.

`hasattr` is slower than `try` and `except` when the common case is that the attribute exists. I recommend not using `hasattr` here.
```bash
$ python3 -m timeit -s 'class C(object): a = 0
o = C()' 'if hasattr(o, "a"): o.a'
2000000 loops, best of 5: 123 nsec per loop
$ python3 -m timeit -s 'class C(object): a = 0
o = C()' 'try: o.a
except: pass'
10000000 loops, best of 5: 39.5 nsec per loop
```

**Comment by justvanrossum** [comment]
Or consider `getattr(o, "a", None)`, that should also be faster for the common case.

**Comment by dscorbett** [comment]
`getattr` doesn’t seem much faster.
```bash
$ python3 -m timeit -s 'class C(object): a = 0
o = C()' 'a = getattr(o, "a", None); a is None'
2000000 loops, best of 5: 121 nsec per loop
```

**Comment by behdad** [comment]
I agree that only setGlyphOrder() should rebuild inverse mapping and assume it doesn't change.

> The most time-consuming callers of `getGlyphID` in the slower build were [`GlyphID.write`](https://github.com/fonttools/fonttools/blob/3.26.0/Lib/fontTools/ttLib/tables/otConverters.py#L281), [`Coverage.preWrite`](https://github.com/fonttools/fonttools/blob/3.26.0/Lib/fontTools/ttLib/tables/otTables.py#L487), and [`SingleSubst.preWrite`](https://github.com/fonttools/fonttools/blob/3.26.0/Lib/fontTools/ttLib/tables/otTables.py#L648).

I think we should do the following:
1. Add a getGlyphIDs that maps an array the fastest way possible.  Then use that in `Coverage.preWrite`.  Rewrite `SingleSubst.preWrite` to also use that.
2. Add a writeArray to GlyphID and make the table writer use that if available, like it currently uses readArray if available.

**Comment by behdad** [comment]
@dscorbett Can you test https://github.com/fonttools/fonttools/pull/2398 ?
